Adds a new sample for datachannel

ErikHellman opened the pull request:
- Messaging between two peers
- Uses new ES6 module importing
- Uses lit-element for web components
- async/await instead of promises

If this looks ok, I'd like to move on to refactor the other samples in a similar style.

ErikHellman committed:
Adds a new sample for datachannel

- Messaging between two peers
- Uses new ES6 module importing
- Uses lit-element for web components
- async/await instead of promises

ErikHellman committed:
- Fix CSS errors and .csslintrc for new properties
- Add sample to index page

ErikHellman committed:
Fix misspelled parameter name

ErikHellman commented:
Ping @fippo. I couldn't add you as a reviewer, but you might be interested in this as well.

fippo commented on src/content/datachannel/messaging/index.html:
```diff
@@ -0,0 +1,68 @@
+<!DOCTYPE html>
+<!--
+ *  Copyright (c) 2015 The WebRTC project authors. All Rights Reserved.
```
masquerarding as jansson (who is ooo currently): wrong year, also below

fippo commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,168 @@
+    try {
+      this._root.querySelector('#connectButton').disabled = true;
+      const dataChannelParams = {ordered: false};
+      let lpc = this._localConnection = new RTCPeerConnection(null);
```
new RTCPeerConnection()
-- passing null was just to workaround Chrome being picky about the lack of arguments in Chrome < 51

fippo commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,168 @@
+  async connect(e) {
+    try {
+      this._root.querySelector('#connectButton').disabled = true;
+      const dataChannelParams = {ordered: false};
```
any reason for using unordered datachannels? It will be somewhat hard to show out-of-order delivery using this sample

ErikHellman replied:
Copy paste mistake. Thanks for noticing. :)

fippo commented on src/content/extensions/multipleroutes/src/utils.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable */
```
drive-by? wfm

ErikHellman replied:
This one kept failing on a lot of stuff when I enable the newer JS stuff. I'll enable it again once I get a chance to refactor it as well.

fippo commented on src/content/datachannel/messaging/index.html:
```diff
@@ -0,0 +1,68 @@
+
+    <script src="https://unpkg.com/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
+    <script src="https://webrtc.github.io/adapter/adapter-latest.js"></script>
+    <script src="main.js" type="module"></script>
```
this works in all browsers supported?

ErikHellman replied:
I've been told it does.

fippo replied:
i've tested in firefox esr (52) and that works too so.. i'm surprised!

fippo commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,168 @@
+  _localChannelOpen(event) {
+    console.log(`Local channel open: ${JSON.stringify(event)}`);
+    this._root.querySelector('#sendLocal').disabled = false;
+    this._root.querySelector('#sendRemote').disabled = false;
```
this doesn't listen for the remote channel open event. Now its pretty hard to get this to fail in this sample but still wrong semantics

ErikHellman replied:
True. I'll add the same to the remote as well for clarity.

ErikHellman committed:
- Fix copyright year
- Remove null from constructor call
- Change to ordered: true for datachannel parameters

ErikHellman committed:
Fix lint errors and bad listener registration

ErikHellman committed:
Fix lint errors

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  disconnect(e) {
+    this._root.querySelector('#disconnectButton').disabled = true;
```
this.shadowRoot

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,169 @@
+      this._root.querySelector('#connectButton').disabled = true;
+      const dataChannelParams = {ordered: true};
+      let lpc = this._localConnection = new RTCPeerConnection();
+      lpc.addEventListener('icecandidate', (e) => this._onIceCandidate(e, this._remoteConnection));
```
only one argument... `e => this...` works fine

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,169 @@
+
+  _channelClosed(event) {
+    console.log(`Channel closed: ${JSON.stringify(event)}`);
+    this._root.querySelector('#sendLocal').disabled = true;
```
with lit-html you might want to do these the otehr way around... it have local variable that you change and just make the elements depend on that. ie don't manually modify that dom in multiple goes

ErikHellman committed:
Change _root to shadowRoot

ErikHellman committed:
Use property for disabled state

ErikHellman committed:
Fixed lint error

ErikHellman commented:
Ping @samdutton 

What's your opinion on having this format for the samples? My argument for using polymer is that all real web projects uses some kind of framework, and showing WebRTC in relation to that makes it easier to relate to.

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,180 @@
+      connected: {
+        type: Boolean,
+        notify: true,
+        reflectToAttribute: true
```
Why reflecting to attribute?

kenchris replied:
It means that your MessagingSample (ie <message-sample> can have a connected attribute like:)

`<message-sample connected>`

kenchris replied:
Also do you need `notify`? isnt it not only for callbacks? like observing when it changes from the outside

samdutton commented:
(Hi @ErikHellman — I wrote a reply a couple of days ago, but I had to restart Chrome at some point, and didn't realise my comment hadn't been posted :/.)

> What's your opinion on having this format for the samples? My argument for using polymer is that all real web projects uses some kind of framework, and showing WebRTC in relation to that makes it easier to relate to.

I take you point about making WebRTC easier to relate to, but personally I prefer the barebones approach.

Unless you're demonstrating a complete app (such as appr.tc) I think it's easier to understand what's going on when you can focus on the core code and not be distracted by scaffolding. 

Using a framework also means more to maintain, more to go wrong and more potential for obsolescence. For example the [YouTube Data API samples](https://developers.google.com/youtube/v3/code_samples/javascript) were written using jQuery, which is now generally avoided for new projects. I also suspect that many WebRTC developers may not be familiar with Polymer.

However, if other WebRTC folk feel differently, I'm not *that* bothered :). As long as it's possible to know from the code what's important, I think that's fine.

kenchris commented:
If you look at the patch, I would actually say that the logic is nicely separated from the UI parts - more so that normally and there is very little scaffolding.

If I am to learn WebRTC that seems perfect, as I don't care much about how the UI is updated but more about the Web RTC logic.

This is using lit-html and not really Polymer, so it is basically only used for stamping out the HTML, which would be much more work manually and result in more irrelevant code, IMHO

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  _firstRendered() {
+    this.shadowRoot.querySelector('#connectButton').addEventListener('click', () => this.connect());
```
You can add these in the HTML.

`<button disabled="${connected}" @click={() => this.connect()}>Connect</button>`

kenchris replied:
actually you probably want

`<button ?disabled=${connected} @click={() => this.connect()}>Connect</button>`

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,180 @@
+    try {
+      console.log(`onIceCandidate: ${JSON.stringify(e)}`);
+      // eslint-disable-next-line no-unused-vars
+      const result = await connection.addIceCandidate(e.candidate);
```
not using result

ErikHellman committed:
Fix click listeners

alvestrand commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,163 @@
+  async connect() {
+    try {
+      const dataChannelParams = {ordered: true};
+      let lpc = this._localConnection = new RTCPeerConnection();
```
style: abbreviations are bad. lpc -> localConnection

fippo replied:
we should try to get some consistency. Other samples as well as lots of wpt use pc1 and pc2

ErikHellman replied:
I agree. I'll remove the abbreviations here.

alvestrand commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,163 @@
+    };
+  }
+
+  _render({connected}) {
```
Here's where I get confused about the point of all this frameworking.
First you use a declarative syntax to draw up the document.
Then you use an imperative syntax to declare an object with properties.
Then you use the imperative syntax to push out declarative syntax that seems to be exactly the same declarative syntax you would write up, but with two layers of extra quoting, and putting your declarative syntax inside the imperatives-syntax file rather than keeping it with the rest of the HTML.
What's the point?

(on a more mundane level: what's the rules for when _render gets called and/or re-called, and who sends parameters to it?)

ErikHellman replied:
Regarding the `_render()` function, the API docs are found here: https://github.com/Polymer/lit-element/blob/master/README.md

About the mix of declarative and imperative, I can understand the confusion. However, this is what a lot of the other frameworks, like React, are doing already, so most web developers will be familiar with it. Even on Android we're doing something similar. For me, the most obvious advantage is data-binding.

ErikHellman replied:
Also, @kenchris can probably give a better answer about the design of this framework. :)

alvestrand commented on src/content/extensions/multipleroutes/src/utils.js:
```diff
@@ -10,8 +10,8 @@
-var pn = chrome.privacy.network;
-var pi = null;
+const pn = chrome.privacy.network;
+const pi = chrome.privacy.IPHandlingPolicy;
```
Don't abbreviate. It obscures.

ErikHellman replied:
Agree. :)

alvestrand commented on src/content/extensions/multipleroutes/src/utils.js:
```diff
@@ -10,8 +10,8 @@
 
-var pn = chrome.privacy.network;
-var pi = null;
+const pn = chrome.privacy.network;
```
Was this just pulled along? It seems a bit irrelevant to the main point of the example.

ErikHellman replied:
II need to make a separate PR that cleans up that file. Will fix.

alvestrand commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  _sendLocalMessage() {
+    const message = this.shadowRoot.querySelector('#localOutgoing').value;
```
where did the shadowRoot enter the picture?
Is this too part of the framework?

ErikHellman replied:
`shadowRoot` is part of the whole Web Components thing. Basically, it's how we access the Shadow DOM.

alvestrand reviewed (commented):
As we talked about on VC, I need to understand why this way is considered "better" than the barebones approach. So far, I don't see what's better about it. Can you try to explain that again?

ErikHellman commented:
The framework I've used here is basically just a thin wrapper on top of the standard Web Components APIs (templates, Shadow DOM and custom elements). An example of how it would look like without lit-elements and only using these standard APIs can be seen here (works only in Chrome at the moment); https://codepen.io/ErikHellman/pen/BPVaZX

Since this is standard APIs that are coming to all browsers, my opinion is that samples for web APIs should embrace this way of creating UI as well. Nobody builds a real application the way the samples are designed today (barebone), and it serves no purpose in samples as they lack context. 

From a developer relations perspective this makes more sense. Samples should be contextual, or lots of developers will have a hard time understanding how to use the APIs in their projects.

ErikHellman committed:
Move incoming messages to property

ErikHellman committed:
Fixed bad leftover code

ErikHellman committed:
Temporary eslint fix - wil remove once  eslint is updated

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,164 @@
+class MessagingSample extends LitElement {
+  constructor() {
+    super();
+    this._localConnection = this._remoteConnection = this._localChannel = this._remoteChannel = null;
```
If you never check if these are null (they are undefined by default) why do this?

kenchris replied:
Anyway they look nice on each on a separate line

kenchris commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,164 @@
+
+      this._remoteConnection.addEventListener('datachannel', e => this._onRemoteDataChannel(e));
+
+      const localOffer = await this._localConnection.createOffer();
```
Could you do the remoteOffer in parallel instead of waiting? ie 

```
const initLocalOffer = async () => {
   ...
}

const initRemoteOffer = async () => {
   ...
}

await Promise.all([initLocalOffer, initRemoteOffer);
console.log("All initialized!")
```

ErikHellman replied:
Tested this and that doesn't work. The result from `createOffer()` must be set before we can call `createAnswer()`.

kenchris replied:
Then do

```
const initLocalOffer = async () => {
   ...
}

const initRemoteOffer = async () => {
   ...
}

await initLocalOffer();
await initRemoveOffer();
```

At least that makes it clear what code belongs where

ErikHellman committed:
Minor cleanups

fippo commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,163 @@
+      await this._localConnection.setLocalDescription(localOffer);
+      await this._remoteConnection.setRemoteDescription(localOffer);
+
+      const remoteOffer = await this._remoteConnection.createAnswer();
```
s/Offer/Answer/

ErikHellman committed:
Rename variable

ErikHellman committed:
Clarify async function calls

ErikHellman committed:
Oh how I love eslint.

ErikHellman commented:
@samdutton @alvestrand 
What is your opinion on this approach? The solution will be similar to what many other framework uses and this will still be based on something that is available as Web APIs in all browsers.

henbos commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,164 @@
+  }
+}
+
+customElements.define('messaging-sample', MessagingSample);
```
nit: newline?

henbos commented on src/content/datachannel/messaging/main.js:
```diff
@@ -0,0 +1,164 @@
+
+  _onRemoteMessageReceived(event) {
+    console.log(`Local message received by remote: ${event.data}`);
+    this.remoteMessages += event.data + '\n';
```
Can we have a unittest that clicks the button and verifies the message is received?

ErikHellman replied:
I will add a UI test once we've switched to Nightwatch. Don't want to write a test using the old fw before then.

samdutton commented:
> Nobody builds a real application the way the samples are designed today (barebone)

That's been deliberate: the samples have been designed to show off WebRTC APIs and methods with as little possible  other code  (within reason!)

> Samples should be contextual, or lots of developers will have a hard time understanding how to use the APIs in their projects.

Ideally, I think you have both: barebones code (to enable devs to understand core APIs and methods) and contextualised code (like appr.tc).

However, as @kenchris says, I don't think it matters that much — and contextualising code does have advantages.

henbos commented:
> The framework I've used here is basically just a thin wrapper on top of the standard Web Components APIs (templates, Shadow DOM and custom elements). An example of how it would look like without lit-elements and only using these standard APIs can be seen here (works only in Chrome at the moment); https://codepen.io/ErikHellman/pen/BPVaZX
> 
> Since this is standard APIs that are coming to all browsers, my opinion is that samples for web APIs should embrace this way of creating UI as well. Nobody builds a real application the way the samples are designed today (barebone), and it serves no purpose in samples as they lack context.
> 
> From a developer relations perspective this makes more sense. Samples should be contextual, or lots of developers will have a hard time understanding how to use the APIs in their projects.

I'm afraid I'm lacking the web guru expertise to have an opinion about Lit since I've never heard of it (but your argument sounds reasonable), I can only serve as an RTCPeerConnection expert in my reviewing. If you need more reviewer's opinions we need to find other reviewers than me since my experience with JavaScript is mostly writing tests for Chrome (using https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/resources/testharness.js).

ErikHellman committed:
Merge remote-tracking branch 'origin/gh-pages' into feature/new-datachannel-sample

# Conflicts:
#	.csslintrc
#	.eslintrc
#	index.html
#	src/content/extensions/multipleroutes/src/utils.js

ErikHellman committed:
Updates after review

ErikHellman committed:
lint fixes
